# EJERCICIOS SIMPLES CON NUMPY 

In [1]:
#comenzamos por importar las librerias:
import numpy as np

1. Tenemos un array de una dimensión con la función range y que contenga 10. Queremos realizar los siguientes ejercicios

In [2]:
x = np.arange(0,10)
x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

a) Mostrar los 5 primeros elementos

In [3]:
x[0:6] #pongo 0 para establecer un rango, aunque también puedo usar :6

array([0, 1, 2, 3, 4, 5])

In [4]:
x[:6]

array([0, 1, 2, 3, 4, 5])

b) Mostrar elementos después del índice 5

In [5]:
x[5:]

array([5, 6, 7, 8, 9])

c) Cada dos elementos, comenzando en el índice 1

In [6]:
x[1::2] #le opngo :: para dercirle que vaya de dos en dos

array([1, 3, 5, 7, 9])

d) Forma de invertir una matriz

In [7]:
x[::-1] #es un slicing o rebanada


array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

e) Mostrar un elemento si y uno no

In [8]:
x[::2]

array([0, 2, 4, 6, 8])

f) Mostrar matriz intermedia que vaya desde la posición 4 hasta la 7

In [9]:
x[4:8]

array([4, 5, 6, 7])

g) Mostrar la matriz invertida cada dos elementos desde el índice 5

In [10]:
x[5::-1]

array([5, 4, 3, 2, 1, 0])

2. Tenemos un array de dos dimensiones y queremos que nos muestre:

In [11]:

x2d=np.random.randint(10, size=(3,4))

In [12]:
x2d

array([[4, 5, 0, 3],
       [0, 7, 5, 2],
       [9, 1, 5, 9]], dtype=int32)

a) dos filas, tres columnas

In [13]:
x2d[1:3, 0:3]

array([[0, 7, 5],
       [9, 1, 5]], dtype=int32)

b) Todas las filas, cada dos columnas

In [14]:
x2d[:,::2]

array([[4, 0],
       [0, 5],
       [9, 5]], dtype=int32)

3.Filtrar números impares: Dado un array 2D, escribe un programa que filtre y devuelva todos los números impares.

In [15]:
# Crear un array 2D
array = np.array([[10,15,20],[25,30,35]])
array

array([[10, 15, 20],
       [25, 30, 35]])

In [16]:
# Filtrar números impares:
impares = array[array %2 !=0] #decimos que si se divide entre 2, el resto sea diferente a 0
impares


array([15, 25, 35])

4.Filtrar números mayores a la media del array: Crea un programa que encuentre y devuelva todos los elementos mayores al promedio del array.

In [17]:
array = np.array([50,50,60,80,90,9090,50])
array

array([  50,   50,   60,   80,   90, 9090,   50])

In [18]:
# Calcular la media
media = array.mean()
media

np.float64(1352.857142857143)

In [19]:
# Filtrar elementos mayores a la media. Construimos una máscara y le pasamos la variable media

mayores = array[array > media]
mayores


array([9090])

In [20]:
#tamabién lo puedo hacer con booleanos
mayores1 =  [array > media]
mayores1

[array([False, False, False, False, False,  True, False])]

5.Suma de arrays y escalar: Dado un array de dimensiones 3×3 suma otro array de las mismas dimensiones y multiplica el resultado por 2.

In [21]:
a = np.array([[1,7,9],[8,10,25],[1,7,9]])
b = np.array([[1,7,6],[8,10,45],[1,7,91]])

In [22]:
# Sumar los arrays y multiplicar por 2






6.Elevar un array al cuadrado y sumar filas: Crea un programa que eleve al cuadrado todos los elementos de un array y luego calcule la suma de cada fila.

In [23]:
# Crear un array



In [24]:
# Elevar al cuadrado





In [25]:
#suma de filas


In [26]:
# Sumar total



7.Sistemas de Ecuaciones:
Resolver un sistema de ecuaciones lineales
Dado el siguiente sistema de ecuaciones:


3x+2y=12

x−y=1

In [27]:
from numpy.linalg import solve
# Matriz de coeficientes
a=np.array([[3,2],[1,-1]])

# Vector de términos independientes
b=np.array([12,1])


In [28]:
# Resolver el sistema
solucion = solve (a,b)
solucion

array([2.8, 1.8])

8.Resolver un sistema de tres ecuaciones
Resuelve el siguiente sistema de ecuaciones:


x+y+z=6

2x+y−z=3

x−y+2z=7

In [29]:
# Matriz de coeficientes
from numpy.linalg import solve
# Matriz de coeficientes
a = np.array([[1,1,1,],[2,1,-1],[1,-1,2]])
b = np.array([6,3,7])

# Vector de términos independientes
soluion=solve(a,b)
solucion



array([2.8, 1.8])

In [30]:
# Resolver el sistema


# -------------------------------------------------------------------------------

# Ejemplo: ¿Cuál es la altura promedio de los presidentes de Estados Unidos?
Los array disponibles en NumPy pueden resultar extremadamente útiles para resumir un conjunto de valores.
Como ejemplo sencillo, consideremos las alturas de todos los presidentes de Estados Unidos.
Estos datos están disponibles en el archivo *president_heights.csv*, que es una lista simple de etiquetas y valores separados por comas:

In [31]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np


In [32]:
# Creamos una variable donde leemos el archivo en este caso formato csv
df = pd.read_csv('president_heights.csv')
df.sample(5)

order                  name  height(cm)
32     35       John F. Kennedy         183
31     34  Dwight D. Eisenhower         179
6       7        Andrew Jackson         185
2       3      Thomas Jefferson         189
36     39          Jimmy Carter         177

In [33]:
# Tomamos de la tabla la columna de height(cm) y la convertimos en un array
altura = np.array(df['height(cm)'])
altura

array([189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175,
       178, 183, 193, 178, 173, 174, 183, 183, 168, 170, 178, 182, 180,
       183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177, 185, 188,
       188, 182, 185])

Ahora que tenemos esta matriz de datos, podemos calcular una variedad de estadísticas resumidas:

In [34]:
#Altura media
media = round(altura.mean(),2)
media

np.float64(179.74)

In [35]:
max = round(altura.max(),2)
max

np.int64(193)

In [36]:
min=round(altura.min(),2)
min


np.int64(163)

In [37]:
std=round(altura.std(),2)
std


np.float64(6.93)

Otros cálculos que podermos realizar:

In [38]:
#percentil 25
per_25 = np.percentile(altura,25)
per_25


np.float64(174.25)

In [39]:
per_75 = np.percentile(altura,75)
per_75


np.float64(183.0)

In [40]:
mediana = np.median(altura)
mediana

np.float64(182.0)


Vemos que la altura media, redondeando, de los presidentes estadounidenses es de 182 cm.

Por supuesto, a veces es más útil ver una representación visual de estos datos, lo que podemos lograr usando herramientas en Matplotlib. 

Por ejemplo, este código genera el siguiente gráfico:

In [41]:
# Importamos la librería matplotlib utilizada para la generación de gráficos




¿Qué tan selecto es el club de presidentes que superan los 185 cm de altura?

Calcula cuántos presidentes pertenecen a este "club" y cuál es su altura media.

In [42]:
masaltos = np.array(df[df['height(cm)']>185]) #Opcion 1
masaltos

array([[1, 'George Washington', 189],
       [3, 'Thomas Jefferson', 189],
       [16, 'Abraham Lincoln', 193],
       [32, 'Franklin D. Roosevelt', 188],
       [36, 'Lyndon B. Johnson', 193],
       [41, 'George H. W. Bush', 188],
       [42, 'Bill Clinton', 188]], dtype=object)

In [43]:
masaltos2 = df['height(cm)'].values  #opcion 2 trayendo solo los valores
masaltos2 = masaltos2[masaltos2>185]
masaltos2

array([189, 189, 193, 188, 193, 188, 188])

In [44]:
masaltos2.mean()

np.float64(189.71428571428572)

Se dice que los primeros presidentes eran más altos.
Calcula la altura media de los primeros 10 presidentes y compárala con los últimos 10.

In [45]:
todas_alturas = df['height(cm)'].values
todas_alturas


array([189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175,
       178, 183, 193, 178, 173, 174, 183, 183, 168, 170, 178, 182, 180,
       183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177, 185, 188,
       188, 182, 185])

In [46]:
primeros_10 = todas_alturas[:10]
primeros_10.mean()

np.float64(177.4)

In [47]:
ultimos_10 = todas_alturas[-10:]
ultimos_10.mean()

np.float64(184.6)

¿Quién está más cerca de la media?

Usando NumPy, encuentra al presidente cuya altura está más cerca de la altura media de todos.

In [52]:
media_altura = np.mean(todas_alturas)
media_altura


np.float64(179.73809523809524)

In [49]:
idx_cercano = np.argmin(np.abs(todas_alturas - media_altura)) #el argumento mas cercano
idx_cercano

np.int64(25)

In [ ]:
presidente_cercano = df.iloc[idx_cercano]['name'] #Iloc me devuelve l posicion que le indico y lo que pongo después es la columna que quiero que me muestre solamente
presidente_cercano#devuelve la posicion iloc

'Woodrow Wilson'

In [51]:
print(f'Altura media es: {media_altura:.2f} cm') #.2f se usa par que coja dos decimales sin redondear 
print(f'El presidente más cercano a la media es {presidente_cercano}')

Altura media es: 179.74 cm
El presidente más cercano a la media es Woodrow Wilson


Ahora usando el Dataset de votos debemo unificar la informacion de ambos DF

In [55]:
df_votos=pd.read_csv('Votos_Populares_Simulados.csv')
df_votos.sample(5)


name  votos_millones
12   Millard Fillmore              20
9          John Tyler              67
38  George H. W. Bush              36
6      Andrew Jackson              70
2    Thomas Jefferson              38

Ahora lo que vamos a hacer es unificar los dos ficheros en un solo DF

In [78]:
df_combinado = df.merge(df_votos, on='name', how='left') #El punto de unión es name que coincide en ambas. Es similar a un left join

In [83]:
df_combinado.sample(5)

order               name  height(cm)  votos_millones
21     23  Benjamin Harrison         168              62
27     30    Calvin Coolidge         178              11
10     11      James K. Polk         173              28
25     28     Woodrow Wilson         180              39
39     42       Bill Clinton         188              68

si por ejemplo las columnas fueran iguales pero con diferente nombre de columnas sería algo así

In [ ]:
df_combinado = df.merge(df_votos, left_on='name', right_on='nombre',how='left')

¿Quién saca más partido de su altura?

Calcula cuántos millones de votos recibe cada presidente por centímetro de altura.
¿Quién es el más "eficiente"? ¿Y el menos?

In [85]:
df_combinado['votos_por_cm'] = df_combinado['votos_millones']/df_combinado['height(cm)'] #creo una unaeva columna que sea de votos por cm
df_combinado.head()

order               name  height(cm)  votos_millones  votos_por_cm
0      1  George Washington         189              48      0.253968
1      2         John Adams         170              61      0.358824
2      3   Thomas Jefferson         189              38      0.201058
3      4      James Madison         163              24      0.147239
4      5       James Monroe         183              52      0.284153

In [87]:
print(df_combinado['votos_por_cm'].idxmax())

34


In [88]:
mas_eficiente=df_combinado.loc[df_combinado['votos_por_cm'].idxmax()]
mas_eficiente

order                        37
name              Richard Nixon
height(cm)                  182
votos_millones               70
votos_por_cm           0.384615
Name: 34, dtype: object

Calcula la correlación entre altura y votos usando NumPy. ¿Existe alguna relación?

In [89]:
#creamos una matriz de correlacion
altura_valores = df_combinado['height(cm)'].values
votos_valores = df_combinado['votos_millones'].values

In [91]:
correlacion = np.corrcoef(altura_valores,votos_valores)[0,1] #es una matriz de una dimension
correlacion

np.float64(0.31235448967212154)

In [92]:
print(f'La correlacion altura-votos es {correlacion}')

La correlacion altura-votos es 0.31235448967212154


¿Qué tan desigual es la distribución de votos?

Calcula el coeficiente de variación (CV) de los votos

In [93]:
stf=np.std(votos_valores)
mean = np.mean(votos_valores)

In [94]:
cv = std / mean
cv

np.float64(0.16698795180722892)

El top 5 presidencial

¿Quiénes son los 5 presidentes con más votos totales?
¿Cuáles son sus alturas? ¿Hay alguna tendencia?

In [102]:
top_5 = df_combinado.sort_values('votos_millones', ascending= False).head(5)
top_5

order             name  height(cm)  votos_millones  votos_por_cm
6       7   Andrew Jackson         185              70      0.378378
34     37    Richard Nixon         182              70      0.384615
28     31   Herbert Hoover         182              69      0.379121
39     42     Bill Clinton         188              68      0.361702
32     35  John F. Kennedy         183              67      0.366120